// type-coverage-core/dist/cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readCache = exports.saveCache = exports.getFileHash = void 0;
const path = require("path");
const fs = require("fs");
const util_1 = require("util");
const crypto_1 = require("crypto");
const readFileAsync = (0, util_1.promisify)(fs.readFile);
const writeFileAsync = (0, util_1.promisify)(fs.writeFile);
const mkdirAsync = (0, util_1.promisify)(fs.mkdir);
async function getFileHash(file, enableCache) {
    return enableCache ? calculateHash((await readFileAsync(file)).toString()) : '';
}
exports.getFileHash = getFileHash;
function calculateHash(str) {
    return (0, crypto_1.createHash)('sha1').update(str).digest('hex');
}
async function saveCache(typeCheckResult, dirName = defaultDirName) {
    await mkdirIfmissing(dirName);
    await writeFileAsync(path.resolve(dirName, 'result.json'), JSON.stringify(typeCheckResult, null, 2));
}
exports.saveCache = saveCache;
const defaultDirName = '.type-coverage';
function statAsync(p) {
    return new Promise((resolve) => {
        fs.stat(p, (err, stats) => {
            if (err) {
                resolve(undefined);
            }
            else {
                resolve(stats);
            }
        });
    });
}
async function mkdirIfmissing(dirName = defaultDirName) {
    const stats = await statAsync(dirName);
    if (!stats) {
        await mkdirAsync(dirName, { recursive: true });
    }
}
async function readCache(enableCache, dirName = defaultDirName) {
    if (!enableCache) {
        return {
            cache: {}
        };
    }
    const filepath = path.resolve(dirName, 'result.json');
    const stats = await statAsync(filepath);
    if (stats && stats.isFile()) {
        const text = (await readFileAsync(filepath)).toString();
        const typeCheckResult = JSON.parse(text);
        if (typeCheckResult && Array.isArray(typeCheckResult.cache)) {
            typeCheckResult.cache = {};
        }
        return typeCheckResult;
    }
    return {
        cache: {}
    };
}
exports.readCache = readCache;
