// type-coverage-core/dist/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lintSync = exports.lint = void 0;
const ts = require("typescript");
const path = require("path");
const minimatch_1 = require("minimatch");
const tsconfig_1 = require("./tsconfig");
const checker_1 = require("./checker");
const dependencies_1 = require("./dependencies");
const ignore_1 = require("./ignore");
const cache_1 = require("./cache");
/**
 * @public
 */
async function lint(project, options) {
    const lintOptions = { ...defaultLintOptions, ...options };
    const { rootNames, compilerOptions } = await (0, tsconfig_1.getProjectRootNamesAndCompilerOptions)(project);
    const program = ts.createProgram(rootNames, compilerOptions, undefined, lintOptions.oldProgram);
    const checker = program.getTypeChecker();
    const allFiles = new Set();
    const sourceFileInfos = [];
    const typeCheckResult = await (0, cache_1.readCache)(lintOptions.enableCache, lintOptions.cacheDirectory);
    const ignoreFileGlobs = lintOptions.ignoreFiles
        ? (typeof lintOptions.ignoreFiles === 'string'
            ? [lintOptions.ignoreFiles]
            : lintOptions.ignoreFiles)
        : undefined;
    for (const sourceFile of program.getSourceFiles()) {
        let file = sourceFile.fileName;
        if (!file.includes('node_modules')) {
            if (!lintOptions.absolutePath) {
                file = path.relative(process.cwd(), file);
                if (!lintOptions.notOnlyInCWD && file.startsWith('..')) {
                    continue;
                }
            }
            if (lintOptions.files && !lintOptions.files.includes(file)) {
                continue;
            }
            if (ignoreFileGlobs && ignoreFileGlobs.some((f) => (0, minimatch_1.minimatch)(file, f))) {
                continue;
            }
            allFiles.add(file);
            const hash = await (0, cache_1.getFileHash)(file, lintOptions.enableCache);
            const cache = typeCheckResult.cache[file];
            if (cache) {
                if (lintOptions.ignoreNested) {
                    cache.anys = cache.anys.filter((c) => c.kind !== 2 /* FileAnyInfoKind.containsAny */);
                }
                if (lintOptions.ignoreAsAssertion) {
                    cache.anys = cache.anys.filter((c) => c.kind !== 3 /* FileAnyInfoKind.unsafeAs */);
                }
                if (lintOptions.ignoreTypeAssertion) {
                    cache.anys = cache.anys.filter((c) => c.kind !== 4 /* FileAnyInfoKind.unsafeTypeAssertion */);
                }
                if (lintOptions.ignoreNonNullAssertion) {
                    cache.anys = cache.anys.filter((c) => c.kind !== 5 /* FileAnyInfoKind.unsafeNonNull */);
                }
            }
            sourceFileInfos.push({
                file,
                sourceFile,
                hash,
                cache: cache && cache.hash === hash ? cache : undefined
            });
        }
    }
    if (lintOptions.enableCache) {
        const dependencies = (0, dependencies_1.collectDependencies)(sourceFileInfos, allFiles);
        for (const sourceFileInfo of sourceFileInfos) {
            if (!sourceFileInfo.cache) {
                (0, dependencies_1.clearCacheOfDependencies)(sourceFileInfo, dependencies, sourceFileInfos);
            }
        }
    }
    let correctCount = 0;
    let totalCount = 0;
    const anys = [];
    const fileCounts = new Map();
    for (const { sourceFile, file, hash, cache } of sourceFileInfos) {
        if (cache) {
            correctCount += cache.correctCount;
            totalCount += cache.totalCount;
            anys.push(...cache.anys.map((a) => ({ file, ...a })));
            if (lintOptions.fileCounts) {
                fileCounts.set(file, {
                    correctCount: cache.correctCount,
                    totalCount: cache.totalCount,
                });
            }
            continue;
        }
        const ingoreMap = (0, ignore_1.collectIgnoreMap)(sourceFile, file);
        const context = {
            file,
            sourceFile,
            typeCheckResult: {
                correctCount: 0,
                totalCount: 0,
                anys: []
            },
            ignoreCatch: lintOptions.ignoreCatch,
            ignoreUnreadAnys: lintOptions.ignoreUnreadAnys,
            catchVariables: {},
            debug: lintOptions.debug,
            strict: lintOptions.strict,
            processAny: lintOptions.processAny,
            checker,
            ingoreMap,
            ignoreNested: lintOptions.ignoreNested,
            ignoreAsAssertion: lintOptions.ignoreAsAssertion,
            ignoreTypeAssertion: lintOptions.ignoreTypeAssertion,
            ignoreNonNullAssertion: lintOptions.ignoreNonNullAssertion,
            ignoreObject: lintOptions.ignoreObject,
            ignoreEmptyType: lintOptions.ignoreEmptyType,
        };
        if (lintOptions.reportSemanticError) {
            const diagnostics = program.getSemanticDiagnostics(sourceFile);
            for (const diagnostic of diagnostics) {
                if (diagnostic.start !== undefined) {
                    totalCount++;
                    let text;
                    if (typeof diagnostic.messageText === 'string') {
                        text = diagnostic.messageText;
                    }
                    else {
                        text = diagnostic.messageText.messageText;
                    }
                    const { line, character } = ts.getLineAndCharacterOfPosition(sourceFile, diagnostic.start);
                    anys.push({
                        line,
                        character,
                        text,
                        kind: 6 /* FileAnyInfoKind.semanticError */,
                        file,
                    });
                }
            }
        }
        sourceFile.forEachChild(node => {
            (0, checker_1.checkNode)(node, context);
        });
        correctCount += context.typeCheckResult.correctCount;
        totalCount += context.typeCheckResult.totalCount;
        anys.push(...context.typeCheckResult.anys.map((a) => ({ file, ...a })));
        if (lintOptions.fileCounts) {
            fileCounts.set(file, {
                correctCount: context.typeCheckResult.correctCount,
                totalCount: context.typeCheckResult.totalCount
            });
        }
        if (lintOptions.enableCache) {
            const resultCache = typeCheckResult.cache[file];
            if (resultCache) {
                resultCache.hash = hash;
                resultCache.correctCount = context.typeCheckResult.correctCount;
                resultCache.totalCount = context.typeCheckResult.totalCount;
                resultCache.anys = context.typeCheckResult.anys;
            }
            else {
                typeCheckResult.cache[file] = {
                    hash,
                    ...context.typeCheckResult
                };
            }
        }
    }
    if (lintOptions.enableCache) {
        await (0, cache_1.saveCache)(typeCheckResult, lintOptions.cacheDirectory);
    }
    return { correctCount, totalCount, anys, program, fileCounts };
}
exports.lint = lint;
const defaultLintOptions = {
    debug: false,
    files: undefined,
    oldProgram: undefined,
    strict: false,
    enableCache: false,
    ignoreCatch: false,
    ignoreFiles: undefined,
    ignoreUnreadAnys: false,
    fileCounts: false,
    ignoreNested: false,
    ignoreAsAssertion: false,
    ignoreTypeAssertion: false,
    ignoreNonNullAssertion: false,
    ignoreObject: false,
    ignoreEmptyType: false,
    reportSemanticError: false,
};
/**
 * @public
 */
function lintSync(compilerOptions, rootNames, options) {
    const lintOptions = { ...defaultLintOptions, ...options };
    const program = ts.createProgram(rootNames, compilerOptions, undefined, lintOptions.oldProgram);
    const checker = program.getTypeChecker();
    const allFiles = new Set();
    const sourceFileInfos = [];
    const ignoreFileGlobs = lintOptions.ignoreFiles
        ? (typeof lintOptions.ignoreFiles === 'string'
            ? [lintOptions.ignoreFiles]
            : lintOptions.ignoreFiles)
        : undefined;
    for (const sourceFile of program.getSourceFiles()) {
        let file = sourceFile.fileName;
        if (!file.includes('node_modules') && (!lintOptions.files || lintOptions.files.includes(file))) {
            if (!lintOptions.absolutePath) {
                file = path.relative(process.cwd(), file);
                if (!lintOptions.notOnlyInCWD && file.startsWith('..')) {
                    continue;
                }
            }
            if (ignoreFileGlobs && ignoreFileGlobs.some((f) => (0, minimatch_1.minimatch)(file, f))) {
                continue;
            }
            allFiles.add(file);
            sourceFileInfos.push({
                file,
                sourceFile,
            });
        }
    }
    let correctCount = 0;
    let totalCount = 0;
    const anys = [];
    const fileCounts = new Map();
    for (const { sourceFile, file } of sourceFileInfos) {
        const ingoreMap = (0, ignore_1.collectIgnoreMap)(sourceFile, file);
        const context = {
            file,
            sourceFile,
            typeCheckResult: {
                correctCount: 0,
                totalCount: 0,
                anys: []
            },
            ignoreCatch: lintOptions.ignoreCatch,
            ignoreUnreadAnys: lintOptions.ignoreUnreadAnys,
            catchVariables: {},
            debug: lintOptions.debug,
            strict: lintOptions.strict,
            processAny: lintOptions.processAny,
            checker,
            ingoreMap,
            ignoreNested: lintOptions.ignoreNested,
            ignoreAsAssertion: lintOptions.ignoreAsAssertion,
            ignoreTypeAssertion: lintOptions.ignoreTypeAssertion,
            ignoreNonNullAssertion: lintOptions.ignoreNonNullAssertion,
            ignoreObject: lintOptions.ignoreObject,
            ignoreEmptyType: lintOptions.ignoreEmptyType,
        };
        sourceFile.forEachChild(node => {
            (0, checker_1.checkNode)(node, context);
        });
        correctCount += context.typeCheckResult.correctCount;
        totalCount += context.typeCheckResult.totalCount;
        anys.push(...context.typeCheckResult.anys.map((a) => ({ file, ...a, sourceFile })));
        if (lintOptions.fileCounts) {
            fileCounts.set(file, {
                correctCount: context.typeCheckResult.correctCount,
                totalCount: context.typeCheckResult.totalCount
            });
        }
    }
    return { correctCount, totalCount, anys, program, fileCounts };
}
exports.lintSync = lintSync;
